### ISSUE TITLE
Language Translation with nn.Transformer and torchtext[BUG] - mask with -inf leads to nans. 

### ISSUE THREAD

**Issue by danielegana** [pytorch/tutorials#2988]
### Add Link

https://pytorch.org/tutorials/beginner/translation_transformer.html

### Describe the bug

Running the tutorial on language translation with transformers leads to nans when training on the first batch iteration on the first epoch, and even when evaluating an untrained model for some input sequences. 

I find this issue simply by copy-pasting the tutorial to my local computer and running. The only line I modify from the tutorial is to replace the DEVICE line for cuda support with a line DEVICE = torch.device('mps' if torch.backends.mps.is_available() else 'cpu') to use mps. 

The issue seems to stem from the target mask. Replacing the line 

mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))

by 

mask = mask.float().masked_fill(mask == 0, float('-1e9')).masked_fill(mask == 1, float(0.0))

allows to train the model for a few batches on the first epoch without nan outputs. A similar problem has been pointed out in https://github.com/pytorch/pytorch/issues/41508

However, even with this "fix", the losses of the model increase with training, and eventually they become nan too. 

### Describe your environment

Running on MacOS with mps support. I am using pytorch 2.2.2 and python 3.9.7.

**Comment by lmntrx-sys** [comment]
**Suggested Steps**
Review Mask Implementation: Investigate the mask implementation in the tutorial and ensure it's correctly applied during training and evaluation.

Gradient Clipping: Implement gradient clipping to prevent exploding gradients, which could lead to NaNs.

Learning Rate: Experiment with different learning rates to see if a lower learning rate stabilizes the training process.

Check Data Preprocessing: Ensure that the input data is correctly preprocessed and normalized

Monitor NaN Values: Add checks to monitor NaN values during training and identify when and where they first appear

**Comment by svekars** [comment]
This tutorial is deprecated. Closing
